packages/lexical/src/__tests__/unit/LexicalNode.test.ts

```diff
@@ -1497,3 +1506,143 @@ describe('LexicalNode tests', () => {
     },
   );
 });
+
+describe('LexicalNode state', () => {
+  initializeUnitTest(
+    (testEnv) => {
+      let root: RootNode;
+
+      beforeEach(async () => {
+        const {editor} = testEnv;
+        await editor.update(() => {
+          root = $getRoot();
+        });
+      });
+
+      test(`setState() need to be inside an update`, async () => {
+        const fn = () => {
+          const keyForString = createStateKey('keyForString', {
+            parse: (value) => value as string,
+          });
+          root.setState(keyForString, 'hello');
+        };
+        expect(fn).toThrow();
+      });
+
+      test(`getState and setState`, async () => {
+        const keyForString = createStateKey('keyForString', {
+          parse: (value) => value as string,
+        });
+        const {editor} = testEnv;
+        editor.update(() => {
+          const stringValue = root.getState(keyForString);
+          type _Test = Expect<Equal<typeof stringValue, string>>;
+          expect(stringValue).toBeUndefined();
+          root.setState(keyForString, 'hello');
+          expect(root.getState(keyForString)).toBe('hello');
+
+          const keyForMaybeString = createStateKey('keyForMaybeString', {
+            parse: (value) => value as string | undefined,
+          });
+          const maybeStringValue = root.getState(keyForMaybeString);
+          type _Test2 = Expect<
+            Equal<typeof maybeStringValue, string | undefined>
+          >;
+          expect(maybeStringValue).toBeUndefined();
+
+          // TO-DO:
+          // It would be a bit nicer if createStateKey also gave a compilation error if parse tried to return null.
+          // It's not a big deal, because getState does give an error. The problem is that I can't activate it.
+          // The @expect-error does not work because strict mode is disabled in `tsconfig.test.json`.
+          const _keyForMaybeNull = createStateKey('keyForMaybeNull', {
+            parse: (value) => value as string | null,
+          });
+          // // @ts-expect-error - null is not a valid value
+          // const _maybeNullValue = root.getState(keyForMaybeNull);
```

Comment:
I don't really understand your comment here, there's no error because you can explicitly cast unknown to anything that you like without any sort of warning or error in typescript. The only difference from `any` is that it requires an explicit "trust me bro". It's perfectly valid to have a function with this type signature, it's just the implementation that's wrong.

```suggestion
          const keyForMaybeNull = createStateKey('keyForMaybeNull', {
            parse: (value) => typeof value === 'string' ? value: null,
          });
          const maybeNullValue = root.getState(keyForMaybeNull);
          expect(maybeNullValue).toBe(null);
```

Reply:
Your intent to restrict the type doesn't really provide any benefit, given `parse: (value) => typeof value === 'string' ? value : null` then `null` will never be serialized if the implementation follows the guidelines that I proposed.

There are use cases where `null` is an explicit value that's different from the default (which could be undefined, or any other value), and your proposed type makes that use case impossible.

Reply:
For more clarity here is a tsplayground demonstrating how I think this should all work https://tinyurl.com/rv8jutvt

<details>
<summary>code below for posterity</summary>

```ts
import { z } from 'zod';
import { Schema, Option } from 'effect';

export type DeepImmutable<T> = T extends Map<infer K, infer V>
  ? ReadonlyMap<DeepImmutable<K>, DeepImmutable<V>>
  : T extends Set<infer S>
  ? ReadonlySet<DeepImmutable<S>>
  : T extends object
  ? {
      readonly [K in keyof T]: DeepImmutable<T[K]>;
    }
  : T;
export type State = {[Key in string]?: StateValue};
// This could be relaxed in a future version, but for now we only
// support values that are directly JSON serializable
export type StateValue = string | number | boolean | null | undefined | State;
export interface StateKey<
  K extends string = string,
  V extends StateValue = StateValue,
> extends StateKeyConfig<V> {
  readonly key: K;
  // Here we are storing a default for convenience
  readonly value: DeepImmutable<V>;
}
export interface StateKeyConfig<V extends StateValue = StateValue> {
  readonly parse: (value?: unknown) => V;
}
export type StateKeyConfigValue<T extends StateKeyConfig> = ReturnType<T['parse']>;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function createStateKey<K extends string, T extends StateKeyConfig<any>>(
  key: K,
  config: T,
): StateKey<K, StateKeyConfigValue<T>> {
  return {key, value: config.parse(undefined), parse: config.parse.bind(config)};
}

// It's best for us to demonstrate actual parsers, but we probably
// shouldn't ship a parse library since many users will already have
// something (zod, effect, whatever)
function parseString<Default extends StateValue>(defaultValue: Default): StateKeyConfig<string | Default> {
  return { parse: (v: unknown) => typeof v === 'string' ? v : defaultValue };
}
const skNull = createStateKey("test", parseString(null));
//    ^?
const skUndefined = createStateKey("test", parseString(undefined));
//    ^?
const skString = createStateKey("test", parseString(''));
//    ^?

const zkNull = createStateKey("test", z.string().nullable().catch(null));
//    ^?
const zkUndefined = createStateKey("test", z.string().optional().catch(undefined));
//    ^?
const zkString = createStateKey("test", z.string().catch(''));
//    ^?

function parseSchema<A extends StateValue, I, Default extends StateValue>(schema: Schema.Schema<A, I, never>, defaultValue: Default): StateKeyConfig<A | Default> {
  return { parse: (value) => Option.getOrElse(Schema.decodeUnknownOption(schema)(value), () => defaultValue) };
}

const ekNull = createStateKey("test", parseSchema(Schema.String, null));
//    ^?
const ekUndefined = createStateKey("test", parseSchema(Schema.String, undefined));
//    ^?
const ekString = createStateKey("test", parseSchema(Schema.String, ''));
//    ^?
```
</details>